LinGUI: Port to GTK4

robxnano opened the pull request:
**Description of Change:**

Port of the Linux client to GTK4 (#5509). Overall, the interface should be mostly very similar to the GTK3 version, but there are some nice improvements that make things look a bit more modern. 

This is a draft for now as I still need to do some proper testing, but I've finished implementing all the features now. At this point I'd like to get some feedback on whether anything else needs to be changed, and also for anyone interested to try out the new version. Ideally I'd like to make sure there aren't any major bugs that I've missed before merging into the main branch, as it could be quite hard for anyone else to work out what I've done due to the huge number of changes which all had to be done at once.

Major changes:
- GTK version 4.4 or later is required. This is available in Ubuntu 22.04, Debian 12, RHEL 9.0 and Fedora 35 or later, so anyone running the latest version of their distribution should be covered. If not, then it's still possible to install the Flatpak version.
- The GStreamer-based preview window has been replaced with one based on the built-in GtkMediaStream. This reduces the number of direct dependencies and will eventually allow for more efficient video playback.
- Header bars are now used for the various secondary windows. With the removal of the GtkToolbar class, I found it was easier to just replace them with header bars, and it also looks better overall. The main window still has the traditional title and menu bar for consistency with HandBrake on other platforms.
- The default audio and subtitle selection options have been moved to their own dialog boxes in order to save space in the main window.
- New custom widgets have been created as GObject subclasses which make use of the property system and make use of object-oriented concepts like encapsulation. This is more in line with modern Gnome applications which I hope will make it easier for new contributors to work on the code in future.

Outstanding issues:
- Theme support: Flatpak only has built-in support for GTK3 themes, so everyone will end up with the generic default theme. There are ways to poke holes in the sandbox to allow access to locally-installed themes, but I don't know whether Flathub will allow them.
- Remove buttons for audio/subtitle tracks: These used an API that was no longer exists, so I had to remove them. I was planning on replacing these GtkTreeView-based lists with a newer alternative, so I haven't replaced the buttons, but for now there's no way to remove individual tracks from the list.
- The GtkMediaStream player doesn't always give good feedback when there is no video codec available to play the video. It should propagate an error that can be handled and displayed in a dialog but that doesn't always happen. There might be something I can do to detect playback failures manually in order to work around this.

**Tested on:**

- [ ] Windows 10+  (via MinGW)
- [ ] macOS 10.13+
- [x] Ubuntu 22.04
- [x] Flatpak
- [x] FreeBSD 14
- [x] NetBSD 9.3
- [x] OpenBSD 7.4

robxnano committed:
LinGUI: Switch build system to GTK4

- Require gtk4 >= 4.4
- Bump glib minimum version to 2.68

robxnano committed:
ci: Update Linux build

- Add libgtk-4-dev and gettext packages
- Remove outdated packages

robxnano committed:
LinGUI: Add GhbButton widget

This widget acts as a replacement for GtkToolButton, providing a
button with both text and icon. It also adds the ability to display
a status indicator on the icon to show the number of items in the
queue.

robxnano committed:
LinGUI: Port to GTK4

- Add accessibility properties
- Add system information to About dialog
- Remove more uses of gtk_dialog_run

robxnano committed:
LinGUI: Remove all legacy GTK 3 code

robxnano committed:
LinGUI: Port to GTK4 (part 2)

robxnano committed:
LinGUI: Remove redundant marshaler generator

robxnano committed:
LinGUI: Add common.h header file

This header contains common header includes and defines which
are used across the GTK GUI code. It is included in all headers
in order to ensure the correct defines are always used.

robxnano committed:
LinGUI: Update and sort headers

- Sort headers according to clang-format rules
- Use #pragma once instead of include guards
- Add G_BEGIN_DECLS around headers

robxnano committed:
LinGUI: Move some items out of signal_user_data_t

robxnano committed:
LinGUI: Remove globals dict

Removes the little-used globals dict and moves its remaining
variables elsewhere.

robxnano committed:
LinGUI: Move builder out of signal_user_data_t

robxnano committed:
LinGUI: Reduce minimum size of main window

robxnano committed:
LinGUI: Move audio/subtitle defaults to separate dialogs

robxnano committed:
LinGUI: Move Add Track buttons to a dropdown menu

Brings the UI into line with the Windows and Mac versions. Also
adds a Clear button to remove all tracks.

robxnano committed:
LinGUI: Use GtkMediaStream for video preview

This API was introduced in GTK4 and handles video playback via
gstreamer or ffmpeg. Removed the gstreamer build dependency as it
is no longer needed.

robxnano committed:
ci: Remove gstreamer package on Linux

sr55 commented:
For the new audio/subtitle behaviour windows.
- Would be nice of these were resizable.
- The add Track button on audio would be better above the grid.  It's a little awkward at the moment.

Looks like there may be a segfault occurring when closing via the file menu.

galad87 commented:
Those are a lot of changes 😅.

Does GtkMediaStream usually have the same decoders as GStreamer, or does it use GStreamer?
I think the theme and GtkMediaStream can be improved later, the button for removing tracks it's a bit more essential 😆

One issue I found: the "Play" button for the live preview doesn't change to a stop button when pressed, and if it's pressed again it begins to encode yet another preview. It would be nice if the hud with the progress bar would stay visible while the preview is being encoded.

robxnano committed:
LinGUI: Make queue context menu more reliable

robxnano committed:
LinGUI: Audio selection dialog fixes

- Move audio selection add button out of list
- Increase size of list
- Enable bitrate radio button by default

robxnano committed:
LinGUI: Allow lower bitrate values for audio selection

Allows the bitrate to be set as if the mono mixdown was selected
when a higher number of channels is requested. This is because the
source tracks may not have that many channels.

robxnano committed:
LinGUI: Add button to remove selected audio/sub track

This is a temporary solution as the track lists will be replaced
in the future.

robxnano commented:
@sr55 What setup were you testing on when you got the segfault? I can't seem to reproduce it. As far as I know, the window is resizable by default, but I did find that under WSLg the correct resize cursor icons were missing as the `adwaita-icon-theme-full` package wasn't installed. Also, I've moved the Add button out of the track list, and made the default size of the window larger.

@galad87 Yes, GTK uses GStreamer as the video backend by default, so it should be able to play anything the old preview window could. I've added a button for removing tracks, and I'll see if I can get the play button to behave better.

robxnano committed:
LinGUI: Only allow one preview encode at a time

- Show stop icon on preview play button while encode in progress
- Allow preview encoding to be stopped

robxnano committed:
LinGUI: Improve preview code

- Refactor to remove some early returns
- Allow for play button tooltip to change when icon changes

robxnano committed:
LinGUI: Keep preview HUD visible while encoding

- Make HUD fade in and out instead of disappearing abruptly

galad87 commented:
I get a crash when I stop a preview or a job in the queue too soon, when it's still scanning the source.

robxnano commented:
The crash while scanning seems to be in libhb, from the backtrace. Not sure what the underlying problem is.
```
#0  0x000000000057e23d in SetWorkStateInfo (job=0x7fff3c0008e0) at ../libhb/work.c:98
#1  work_func (_work=0x660a780) at ../libhb/work.c:171
#2  0x0000000000540a6b in hb_thread_func (_t=0x4d1c6b0) at ../libhb/ports.c:862
#3  0x00007ffff62d1897 in start_thread () at /lib64/libc.so.6
#4  0x00007ffff635880c in clone3 () at /lib64/libc.so.6
```
https://github.com/HandBrake/HandBrake/blob/0d0c12253117469f32cc078cbc12f5de332e4dd1/libhb/work.c#L98

galad87 commented:
I think it happened before too, so it's not something related to this PR.

galad87 commented:
Anyway, the GTK4 UI seems to work well, If you want to merge this I have nothing against it.

robxnano commented:
Yeah, I think I'm happy with the changes as they are currently. If anyone else wants to take a look that's fine of course.

galad87 commented:
Merged, thanks for all the work.
I updated the workflow on https://github.com/HandBrake/HandBrake-snapshots/ , hopefully I got everything right.
